### ISSUE TITLE
Semantinc error reporting

### ISSUE THREAD

**Issue by soywiz** [TiVo/intellij-haxe#291]
Detecting errors as soon as possible is pretty important because reduces development times (editing -> compiling -> running).
As far as I know currently just are detecting at editing time syntax errors and some errors related to failed resolving class or instance names. But there are a lot of other semantic errors.

For example:
```
var a = 10;
a = "hello";
```

Is syntactically correct, doesn't have problems resolving names, but it is invalid. You cannot reassign a variable that has type int with a string.

Other example:
```
interface Test1 { function a(); }
class Test2 implements Test1 { }
```

Test2 doesn't implements Test1 methods.

Other more example:
```
for (n in 10) { trace(n); }
```

Int doesn't conform the Iterable typedef.

So it would be great to be able to detect those kind of errors.

In order to do that, we need a new phase, that should be executed (if doesn't exist yet) after opening a file or after editing a file, and after the file and dependent files have been parsed so we have context information (can resolve any referenced type and check members).

Do you know how it would be possible to trigger that event in intelliJ? I think I could do the code that checks these semantinc errors, opening a new kind of error reporting so we can improve it bit by bit.

**Comment by soywiz** [comment]
I have found Annotators: https://confluence.jetbrains.com/display/IntelliJIDEA/Annotator

So probably that will be the way to go. It should work fine directly for detecting inheritance issues, but not sure about detecting in-method issues. I will think and will experiment about that.

**Comment by as3boyan** [comment]
Yeah, it's annotators

>Do you know how it would be possible to trigger that event in intelliJ? I think I could do the code that checks these semantinc errors, opening a new kind of error reporting so we can improve it bit by bit.

Of course it's possible, and it's great, I can help you with implementation, it's exciting to me but there are some pretty critical issues there, for example honestly it's hard to do macro stuff and use HaxeFlixel(because of generics) and there are basically tons of things to do, and I am responsible for refactoring features also.

I will write a basic instruction if someone wants to do it

Basically you can implement it yourself if you want, there is great guide on that:
http://www.jetbrains.org/intellij/sdk/docs/tutorials/custom_language_support/annotator.html

Basically copy that annotator from sample, and 

```haxe
interface Test1 { function a(); }
class Test2 implements Test1 { }
```

For this case, you may want to:

1. Iterate through classes
2. Get inheritor of each class, for start you can start with single inheritor, there is helper function for that
3. Get/resolve inheritor, for start do only if it's `instanceof HaxeInterfaceDeclaration` or so
4. There is function to get methods of class/interface, so basically check if methods are implemented.
5. Show annotation (just like in sample)

You can event add quick fix for it like `Implement missing methods` stuff like JetBrains did for Java

**Comment by soywiz** [comment]
I know that preprocessor and macros limit a lot refactoring and makes things harder for IDEs. Some years ago I used D language. It was powerful because of the macros, but compilation times were crazy on a very very large project I did that used a lot of macros and compile-time generation code. Also disallowed to do incremental compilation, and the complexity made a IDE hard. So I had a bad experience with preprocessor. I was using it for generating code, but people is using preprocessor in haxe to implement lack of features: like generators, async, value-object constructors. So I don't like preprocessors and I like more generating code at runtime when it is possible or other ways of doing things.

D at least didn't had preprocessor, but a "static if" http://dlang.org/version.html that didn't break the syntax of the language. But since you can use the Haxe preprocessor everywhere, a generic parsing without knowing the preprocessor declarations is very hard and prone to errors and bugs.

So probably the things I like less from haxe are: preprocessor, macros, no DRY getters, constructors without being able to create fields and lack of generatords and await...async.

---

Regarding to generics I think that we can ride that.

I have already used annotator today.
https://github.com/soywiz/intellij-haxe/commit/77148d5435749097ce511dc04b737866a7de13c8

I have already done a small semantic error reporting. It checks if the package element matches the folder. I was not able to detect where are the source roots so it is not perfect, but at least it is a start.

Since the annotator runs  for each psielement. I am doing thte analysis just once and then displaying the stored errors as userData.

**Comment by soywiz** [comment]
@as3boyan 

I have implemented package fixer. It checks the source content roots to determine the right package:

![fix_package](https://cloud.githubusercontent.com/assets/570848/7990898/730e156a-0af4-11e5-9666-e6bbc01b8977.png)

But I have a problem. I should detect content root of haxe std, and maybe libraries. Is there code to query for those folders?

![fix_package_not_detecting_framework](https://cloud.githubusercontent.com/assets/570848/7990907/860d5f4a-0af4-11e5-9835-b866e67e9d3e.png)

**Comment by as3boyan** [comment]
std is a content root, and if I got it right, then it should be present in the SDK, we used to work with such paths when we worked on `haxelib discovery` feature, which detects used libs from project.xml/hxml/project arguments and adds it to it

**Comment by as3boyan** [comment]
We used to get classpath list for module, https://github.com/TiVo/intellij-haxe/blob/ef4e8880f37269e43d7413bf98427d98853fe56d/src/com/intellij/plugins/haxe/haxelib/HaxelibManager.java#L85
Not sure how to make it for sdk
https://github.com/TiVo/intellij-haxe/issues/17

**Comment by as3boyan** [comment]
there are should be way to get content root

BTW, great work

**Comment by soywiz** [comment]
Regarding to source roots, I have been managed to get the sdk source roots. Probably still require some work for libraries, but at least it is a step forward.

Furthermore to method and field type detection (you can return a constant field and the constant will be propagated) I have implemented override errors.

![override_support](https://cloud.githubusercontent.com/assets/570848/7995653/4f449a18-0b18-11e5-8618-90c7337fba78.png)

**Comment by EBatTiVo** [comment]
There is a file HaxeLibRootDetector which does some of that.  

We also implemented HaxelibClasspahtUtils, which has a number of useful functions that all use the locally installed 'haxelib' executable to find the libraries as the compiler does.  Of particular note are getProjectLibraryClasspath() and getSdkClasspath().

**Comment by soywiz** [comment]
Note for me:

```
class Test extends BaseTest { public function new() { } } // Here
class BaseTest { public function new() { } }
```

Informs wrongly that it requires an override, i have to check if it is the constructor. Also I should enforce a super(); in the case the parent class has a constructor.

With a bit more of work I will be able to check function calls too to ensure signature.

**Comment by soywiz** [comment]
I have fixed that in the last commit and also checked that you are not trying to override static or inline methods (I have to check @:final too) and that you are not changing the visibility when overriding.
I should check that wew don't have things like "public public" or "public private" (that modifiers are not repeated or that you are not using severals of the same category like public/private)

**Comment by soywiz** [comment]
I have added code that checks if a class is implementing an interface

![screen shot 2015-06-05 at 18 03 25](https://cloud.githubusercontent.com/assets/570848/8010053/50bc6d0e-0bad-11e5-97d9-e824275550a8.png)

I want to add a quick fix to implement them. I have seen that there is HaxeImplementMethodHandler, but the code is a bit coupled as I have seen. There is reusable code that with a psielement of a class + psielement of interface methods, I can generate code for those methods?

**Comment by as3boyan** [comment]
Awesome

Code for https://github.com/TiVo/intellij-haxe/blob/ImplementNestedFunctionTypesSupport/src/common/com/intellij/plugins/haxe/ide/generation/OverrideImplementMethodFix.java - may be some parts can be reused, perhaps we can make an util class.

https://github.com/TiVo/intellij-haxe/blob/ImplementNestedFunctionTypesSupport/src/common/com/intellij/plugins/haxe/ide/generation/CreateGetterSetterFix.java Seems to work in a similar way.

https://github.com/TiVo/intellij-haxe/blob/ImplementNestedFunctionTypesSupport/src/common/com/intellij/plugins/haxe/ide/generation/BaseCreateMethodsFix.java#L88-88 - seems to be heavily used for most method generating actions.

`HaxeElementGenerator` class I used a lot when I implemented some refactoring features(push up/push down) basically it used to work in a same way. But I got an issue with formatting, I don't really understand how it works for getter/setter and override/implement methods action, not sure, can't find code for it there, not sure.

You prepare text code, then pass it to HaxeElementGenerator puts it in some sandbox(creates dummy file and parses it I guess), and returns members.

In push up refactoring, we generate method in interface from a class, I used something like this
```java
methodCopy = HaxeElementGenerator.createFunctionPrototypeDeclarationWithAttributes(myProject, methodCopy.getText().trim() + ";");
```
Basically it does same thing - generates method from text.

Looking at `OverrideImplementMethodFix` - there is should be some way to call it or activate, I think.....

**Comment by EBatTiVo** [comment]
> I should check that wew don't have things like "public public" or "public private" (that modifiers are not repeated or that you are not using severals of the same category like public/private)

That can be (and some of it is) enforced in the BNF.

What's a bit more interesting is that visibility rules in Haxe override the stated visibility by the code.  That is, the 'private' keyword is redundant and ignored in most places.  http://haxe.org/manual/class-field-visibility.html.  That is something that can't be enforced in the BNF, but might be useful for developers to know.

**Comment by as3boyan** [comment]
A lil bit offtopic: I got a funny glitch in IDEA http://i.imgur.com/hDId6mv.png

**Comment by soywiz** [comment]
:P I didn't read this. So intelliJ is now owning browsers too.

**Comment by soywiz** [comment]
I'm closing it, since it is partially implemented. And I would create issues for false-positives and for specific errors that are not reported at this point.
